src/Bedrock.Framework/Protocols/MessagePipeReader.cs

```diff
@@ -0,0 +1,179 @@
+﻿using System;
+using System.Buffers;
+using System.IO;
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Bedrock.Framework.Protocols
+{
+    public class MessagePipeReader : PipeReader
+    {
+        private readonly IMessageReader<ReadOnlySequence<byte>> _messageReader;
+        private readonly PipeReader _reader;
+
+        private SequencePosition _examined;
+        private SequencePosition _consumed;
+        private ReadOnlySequence<byte> _message;
+        private bool _isThisCompleted;
+        private bool _isCanceled;
+        private bool _isCompleted;
+        private readonly ConsumableArrayBufferWriter<byte> _backlog = new ConsumableArrayBufferWriter<byte>();
+        private bool _allExamined;
+
+        public MessagePipeReader(PipeReader reader, IMessageReader<ReadOnlySequence<byte>> messageReader)
+        {
+            _reader = reader ?? throw new ArgumentNullException(nameof(reader));
+            _messageReader = messageReader ?? throw new ArgumentNullException(nameof(messageReader));
+        }
+
+        public override void AdvanceTo(SequencePosition consumed)
+        {
+            AdvanceTo(consumed, consumed);
+        }
+
+        public override void AdvanceTo(SequencePosition consumed, SequencePosition examined)
+        {
+            if (_isThisCompleted)
```

Comment:
We have the choice of either checking for completion here, or squeezing out a little more performance by relying on the underlying PipeReaders completion checking.

The disadvantage of the latter is that we are not guaranteed to get the right exception. For example `AdvanceTo` will throw an `ArgumentOutOfRangeException` when it starts enumerating the consumed bytes, since the underlying buffer has been cleared.

Reply:
I would go with the first.

Perf-wise no difference should show up, as 
* the rest of the method has more to do than a simple `cmp`
* branch predictors are pretty good nowadays
* a throw-helper is used, so the JIT will place it at the end of the "method"
* when the branch-predictor has no history for this branch, it will predict "not taken" as the jump to the throw-helper is a forward branch